{
    "addition": {},
    "removed": {
        "1": "/*\n",
        "2": " * Copyright (C) 2007 The Guava Authors\n",
        "3": " *\n",
        "4": " * Licensed under the Apache License, Version 2.0 (the \"License\");\n",
        "5": " * you may not use this file except in compliance with the License.\n",
        "6": " * You may obtain a copy of the License at\n",
        "7": " *\n",
        "8": " * http://www.apache.org/licenses/LICENSE-2.0\n",
        "9": " *\n",
        "10": " * Unless required by applicable law or agreed to in writing, software\n",
        "11": " * distributed under the License is distributed on an \"AS IS\" BASIS,\n",
        "12": " * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
        "13": " * See the License for the specific language governing permissions and\n",
        "14": " * limitations under the License.\n",
        "15": " */\n",
        "16": "\n",
        "17": "package com.google.common.base;\n",
        "18": "\n",
        "19": "import com.google.common.annotations.GwtCompatible;\n",
        "20": "\n",
        "21": "import junit.framework.TestCase;\n",
        "22": "\n",
        "23": "import java.util.Iterator;\n",
        "24": "import java.util.NoSuchElementException;\n",
        "25": "\n",
        "26": "/**\n",
        "27": " * Unit test for {@code AbstractIterator}.\n",
        "28": " *\n",
        "29": " * @author Kevin Bourrillion\n",
        "30": " */\n",
        "31": "@GwtCompatible(emulated = true)\n",
        "32": "// TODO(cpovirk): why is this slow (>1m/test) under GWT when fully optimized?\n",
        "33": "public class AbstractIteratorTest extends TestCase {\n",
        "34": "\n",
        "35": "  public void testDefaultBehaviorOfNextAndHasNext() {\n",
        "36": "\n",
        "37": "    // This sample AbstractIterator returns 0 on the first call, 1 on the\n",
        "38": "    // second, then signals that it's reached the end of the data\n",
        "39": "    Iterator<Integer> iter = new AbstractIterator<Integer>() {\n",
        "40": "      private int rep;\n",
        "41": "      @Override public Integer computeNext() {\n",
        "42": "        switch (rep++) {\n",
        "43": "          case 0:\n",
        "44": "            return 0;\n",
        "45": "          case 1:\n",
        "46": "            return 1;\n",
        "47": "          case 2:\n",
        "48": "            return endOfData();\n",
        "49": "          default:\n",
        "50": "            fail(\"Should not have been invoked again\");\n",
        "51": "            return null;\n",
        "52": "        }\n",
        "53": "      }\n",
        "54": "    };\n",
        "55": "\n",
        "56": "    assertTrue(iter.hasNext());\n",
        "57": "    assertEquals(0, (int) iter.next());\n",
        "58": "\n",
        "59": "    // verify idempotence of hasNext()\n",
        "60": "    assertTrue(iter.hasNext());\n",
        "61": "    assertTrue(iter.hasNext());\n",
        "62": "    assertTrue(iter.hasNext());\n",
        "63": "    assertEquals(1, (int) iter.next());\n",
        "64": "\n",
        "65": "    assertFalse(iter.hasNext());\n",
        "66": "\n",
        "67": "    // Make sure computeNext() doesn't get invoked again\n",
        "68": "    assertFalse(iter.hasNext());\n",
        "69": "\n",
        "70": "    try {\n",
        "71": "      iter.next();\n",
        "72": "      fail(\"no exception thrown\");\n",
        "73": "    } catch (NoSuchElementException expected) {\n",
        "74": "    }\n",
        "75": "  }\n",
        "76": "\n",
        "77": "  public void testSneakyThrow() throws Exception {\n",
        "78": "    Iterator<Integer> iter = new AbstractIterator<Integer>() {\n",
        "79": "      boolean haveBeenCalled;\n",
        "80": "      @Override public Integer computeNext() {\n",
        "81": "        if (haveBeenCalled) {\n",
        "82": "          fail(\"Should not have been called again\");\n",
        "83": "        } else {\n",
        "84": "          haveBeenCalled = true;\n",
        "85": "          sneakyThrow(new SomeCheckedException());\n",
        "86": "        }\n",
        "87": "        return null; // never reached\n",
        "88": "      }\n",
        "89": "    };\n",
        "90": "\n",
        "91": "    // The first time, the sneakily-thrown exception comes out\n",
        "92": "    try {\n",
        "93": "      iter.hasNext();\n",
        "94": "      fail(\"No exception thrown\");\n",
        "95": "    } catch (Exception e) {\n",
        "96": "      if (!(e instanceof SomeCheckedException)) {\n",
        "97": "        throw e;\n",
        "98": "      }\n",
        "99": "    }\n",
        "100": "\n",
        "101": "    // But the second time, AbstractIterator itself throws an ISE\n",
        "102": "    try {\n",
        "103": "      iter.hasNext();\n",
        "104": "      fail(\"No exception thrown\");\n",
        "105": "    } catch (IllegalStateException expected) {\n",
        "106": "    }\n",
        "107": "  }\n",
        "108": "\n",
        "109": "  public void testException() {\n",
        "110": "    final SomeUncheckedException exception = new SomeUncheckedException();\n",
        "111": "    Iterator<Integer> iter = new AbstractIterator<Integer>() {\n",
        "112": "      @Override public Integer computeNext() {\n",
        "113": "        throw exception;\n",
        "114": "      }\n",
        "115": "    };\n",
        "116": "\n",
        "117": "    // It should pass through untouched\n",
        "118": "    try {\n",
        "119": "      iter.hasNext();\n",
        "120": "      fail(\"No exception thrown\");\n",
        "121": "    } catch (SomeUncheckedException e) {\n",
        "122": "      assertSame(exception, e);\n",
        "123": "    }\n",
        "124": "  }\n",
        "125": "\n",
        "126": "  public void testExceptionAfterEndOfData() {\n",
        "127": "    Iterator<Integer> iter = new AbstractIterator<Integer>() {\n",
        "128": "      @Override public Integer computeNext() {\n",
        "129": "        endOfData();\n",
        "130": "        throw new SomeUncheckedException();\n",
        "131": "      }\n",
        "132": "    };\n",
        "133": "    try {\n",
        "134": "      iter.hasNext();\n",
        "135": "      fail(\"No exception thrown\");\n",
        "136": "    } catch (SomeUncheckedException expected) {\n",
        "137": "    }\n",
        "138": "  }\n",
        "139": "\n",
        "140": "  public void testCantRemove() {\n",
        "141": "    Iterator<Integer> iter = new AbstractIterator<Integer>() {\n",
        "142": "      boolean haveBeenCalled;\n",
        "143": "      @Override public Integer computeNext() {\n",
        "144": "        if (haveBeenCalled) {\n",
        "145": "          endOfData();\n",
        "146": "        }\n",
        "147": "        haveBeenCalled = true;\n",
        "148": "        return 0;\n",
        "149": "      }\n",
        "150": "    };\n",
        "151": "\n",
        "152": "    assertEquals(0, (int) iter.next());\n",
        "153": "\n",
        "154": "    try {\n",
        "155": "      iter.remove();\n",
        "156": "      fail(\"No exception thrown\");\n",
        "157": "    } catch (UnsupportedOperationException expected) {\n",
        "158": "    }\n",
        "159": "  }\n",
        "160": "\n",
        "161": "  public void testReentrantHasNext() {\n",
        "162": "    Iterator<Integer> iter = new AbstractIterator<Integer>() {\n",
        "163": "      @Override protected Integer computeNext() {\n",
        "164": "        hasNext();\n",
        "165": "        return null;\n",
        "166": "      }\n",
        "167": "    };\n",
        "168": "    try {\n",
        "169": "      iter.hasNext();\n",
        "170": "      fail();\n",
        "171": "    } catch (IllegalStateException expected) {\n",
        "172": "    }\n",
        "173": "  }\n",
        "174": "\n",
        "175": "  // Technically we should test other reentrant scenarios (4 combinations of\n",
        "176": "  // hasNext/next), but we'll cop out for now, knowing that \n",
        "177": "  // next() both start by invoking hasNext() anyway.\n",
        "178": "\n",
        "179": "  /**\n",
        "180": "   * Throws a undeclared checked exception.\n",
        "181": "   */\n",
        "182": "  private static void sneakyThrow(Throwable t) {\n",
        "183": "    class SneakyThrower<T extends Throwable> {\n",
        "184": "      @SuppressWarnings(\"unchecked\") // intentionally unsafe for test\n",
        "185": "      void throwIt(Throwable t) throws T {\n",
        "186": "        throw (T) t;\n",
        "187": "      }\n",
        "188": "    }\n",
        "189": "    new SneakyThrower<Error>().throwIt(t);\n",
        "190": "  }\n",
        "191": "\n",
        "192": "  private static class SomeCheckedException extends Exception {\n",
        "193": "  }\n",
        "194": "\n",
        "195": "  private static class SomeUncheckedException extends RuntimeException {\n",
        "196": "  }\n",
        "197": "}\n",
        "198": "\n"
    }
}